Elaborate on definition for Conformance tests (#12744)
kubevirt/kubevirt: docs/conformance.md

# Conformance testing

> The Conformance tests are subset of our e2e test suite that defines set of features
> defining Kubevirt core functionality. The goal of the suite is to make sure that selected features work across different Vendors, CNIs, CSIs, CRIs implementations.

Reviewer: ```suggestion
The Conformance tests are a subset of our e2e test suite that cover a set of features
defining Kubevirt core functionalities. The goal of the suite is to make sure that selected features work across different Vendors, CNIs, CSIs, CRIs implementations.
```

## How are Conformance tests selected

By default tests with a `[Conformance]` tag (and from release 0.59 also decorator) will be executed and tests with a
`[Disruptive]` tag will be skipped. These tests will not destroy or modify
anything outside of the explicitly created test namespaces by the binary.

> The tests are always part of Kubevirt e2e tests that are usually approved by individual [Sigs](https://github.com/kubevirt/community/blob/main/sig-list.md#special-interest-groups). For the Conformance tests we have higher requirments than the regular tests, most significant requirments are:

Reviewer: ```suggestion
The tests are always part of Kubevirt e2e tests that are usually approved by individual [Sigs](https://github.com/kubevirt/community/blob/main/sig-list.md#special-interest-groups). For the Conformance tests we have higher requirements than the regular tests, most significant requirements are:
```

- Test is not Disruptive
- Test is not flaky and is reliable per data available from CI team

> - For more concreate details see below

Reviewer: ```suggestion
- For more concrete details see below
```

> ## Conformance tests requirments:

Reviewer: ```suggestion
## Conformance tests requirements:
```

- The feature being tested is GA (no feature gate)
- The feature is not deprecated
- The feature is turned on by default (no feature toogle)
- The test uses only GAed Kubernetes APIs/features

> - The test uses only APIs available to user

Reviewer 2: Should the eventual goal here to also only use GA'd KubeVirt features without FGs? I appreciate that isn't possible right now so the current wording is fine but would be nice to document something here.

Author: See first point "- The feature being tested is GA (no feature gate)"

Reviewer 2: ACK apologies missed that.

Author: Thank you for the review. I can add the reasoning if some points need to.

- The test doesn't use direct access to virt-launcher/libvirt
- It doesn't require any special hardware (such as gpu/pci devices,...) being available on the cluster other than the default required device nodes (kvm, virtio-net)
- It doesn't use `Skip` or any other helper that uses it
- It does not require access to public network, all images or resources should be able to be pre-pulled
- It does not rely on environment of the executor
- It has history of stability and is not flaky, and consequently doesn't have the QUARANTINE label.
- It doesn't rely on specific implementation of CRI, CSI, CNI

Exceptions are allowed but must be explicit and reasoned.

## Delivery

KubeVirt releases a `kubevirt/conformance:<release>` image and
`conformance.yaml` manifest starting from `v0.33.0`. It can be executed as a

> [Sonobuoy](https://sonobuoy.io/) plugin to verify if the underlying cluster

Reviewer 3: Just a question outside the scope of this PR. 

Do we still want to use sonobuoy for executing these conformance tests? I think we see some limitations there like we are currently not running the tests in parallel. 

Thanks for this @xpivarc  - it should be a great help.

Author: I would say we should follow Kubernetes here to give the same experience.

# Conformance testing

The Conformance tests are subset of our e2e test suite that defines set of features
defining Kubevirt core functionality. The goal of the suite is to make sure that selected features work across different Vendors, CNIs, CSIs, CRIs implementations.

## How are Conformance tests selected

> By default tests with a `[Conformance]` tag (and from release 0.59 also decorator) will be executed and tests with a

Reviewer 4: How about adding the label decorator?

Author: What do you mean by adding?

Reviewer 4: I mean, here we are only mentioning the `[Conformance]` tag.I think we want to switch using the Conformance label decorator.

Author: Right, as we don't have the label yet I would leave this out and fill in later.

Reviewer 5: Part of another PR: https://github.com/kubevirt/kubevirt/pull/12682/files#diff-2be6ff8bee7a7b41764cd824e3e7afbc1ae3facd2893bdeb60f7ac978303a387R8

> `[Disruptive]` tag will be skipped. These tests will not destroy or modify

Author: I would keep this as documentation only for now

anything outside of the explicitly created test namespaces by the binary.

> The tests are always part of Kubevirt e2e tests that are usually approved by individual Sigs. For the Conformance tests we require higher requirments than than 

Reviewer 5: ```suggestion
The tests are always part of Kubevirt e2e tests that are usually approved by individual Sigs. For the Conformance tests we have higher requirements than the 
```

> The tests are always part of Kubevirt e2e tests that are usually approved by individual Sigs. For the Conformance tests we require higher requirments than than 

Author: Sounds good

regular tests, most significant requirments are:
- Test is not Disruptive
- Test is not flaky and is reliable per data available from CI team
- For more concreate details see below

## Conformance tests requirments:

> - The feature being tested is GA (no feature gated)

Reviewer 6: ```suggestion
- The feature being tested is GA (no feature gate)
```

- The feature is not deprecated

> - The feature is turned on by default (no feature toogle)

Reviewer 4: An exception to this could be upgrade test. WDYT?

Author: I am not sure upgrade tests are eligible as they do modify the installation. I believe that has to be leave for vendors, users to tests for themself...

Reviewer 4: ```suggestion
- The feature is turned on by default (no feature toggle)
```

- The test uses only GAed Kubernetes APIs/features
- The test uses only APIs available to user
- The test doesn't use direct access to virt-launcher/libvirt
- It doesn't require any special hardware (such as gpu/pci devices,...) being available on the cluster other than the default required device nodes (kvm, virtio-net)
- It doesn't use `Skip` or any other helper that uses it
- It does not require access to public network, all images or resources should be able to be pre-pulled
- It does not rely on environment of the executor

> - It has history of stability and is not flaky

Reviewer 5: nit

```suggestion
- It has history of stability and is not flaky, also it doesn't have the QUARANTINE label.
```
